nbbuild/templates/projectized.xml

```diff
@@ -290,7 +290,7 @@
     <target name="test-lib-init" depends="-init-bootclasspath-prepend,init,-build-libs.junit4">
         <path id="test.unit.lib.cp"/>
         <property name="test.jms.flags" value=""/>
-        <property name="test.run.args" value="-ea -Xmx700m ${metabuild.jms-flags.jvm} ${test.jms.flags} -XX:+IgnoreUnrecognizedVMOptions"/>
+        <property name="test.run.args" value="-ea -Xms1200m -Xmx1200m -XX:+UseParallelGC ${metabuild.jms-flags.jvm} ${test.jms.flags} -XX:+IgnoreUnrecognizedVMOptions"/>
```

Comment:
Everytime I see people mess with the GC I have to ask this: Is there a measurable improvement switching from G1GC to ParallelGC and does it depend on the JVM version? Pinning the heapsize in my mind is ok.

Reply:
Hi Matthias,

ParallelGC is a generational throughput GC. It is available on all JDKs from all hotspot vendors and designed for tasks which are willing to trade latency for throughput (batch processing or tests). The GC can do that since it is not concurrent, and has therefore no additional runtime overhead (for both CPU and heap upkeep overhead).

G1 (new default), Shenandoha and ZGC are the concurrent GCs which try to keep latency low but have more overhead. G1 is general purpose, ZGC is mostly for very large heaps, Shenandoah scales well from small to large heaps and has also lower heap size overhead compared to other concurrent GCs, but is not included in Oracle JDKs. (picking between the 3 concurrent GCs is often application specific and requires testing)

SerialGC is like parallel just single threaded, nice for small heaps, e.g embedded or single threaded micro service deployment (Shenandoah is also nice for embedded if you have multiple cores and low throughput).

Our nodes (both gh actions and travis have 2 threads minimum, and 7GB RAM (travis might have 4GB)) so we can use ParallelGC instead of SerialGC. Further, increasing heap size even allows to run some tests without triggering a GC cycle at all. By using a non-concurrent GC we also get lower overhead, even if it doesn't run. We could likely further tweak this by increasing Xmx, but the gain might be fairly small.

Did I measure that for the tests? No since we have 14h of test jobs. But since we are looking for throughput and don't care about latency the choice is fairly simple since there is only one GC available on hotspot distributions which is made for this usecase.

Reply:
we can measure. We can compare average (non-php) travis runs from after this PR to before if there are any changes. What I mean is that it is annoying to test it locally since we have so many hours of tests.

Conceptionally, ParallelGC is the right GC for this use case since it is the only "throughput GC" there is. Setting a GC has also the benefit of getting predictable results. Tests on 8 would use ParallelGC but anything beyond 8 would likely use G1GC. I say likely since there is also ergonomics code in the JVM, If it does not detect enough cores or memory it might pick something else - this is not specified and could change from version to version, how well the container detection works etc. So setting a GC for tests gives us the opposite of noise IMO.

Serial and Parallel GCs fill an important niche and are still getting new features. I use ParallelGC for example for the maven daemon too. It has less memory and less computational overhead compared to the concurrent GCs which is exactly what you want for worker processes.

Reply:
in another project we use EpsilonGC for tests - it is the most predictable and fastest GC since it doesn't run at all, the JVM exits if it runs out of memory ;)

Reply:
Noise was not meant in measurement gitter, but unnecessary configuration. I suspect changing the GC will not impact the test runtime only minimally and thus the configuration should not be there. Anyway, I made my point, so lets stop discussing and keep as is.